dask/async.py

```diff
@@ -458,16 +463,28 @@ def fire_task():
             data = dict((dep, state['cache'][dep])
                         for dep in get_dependencies(dsk, key))
             # Submit
-            apply_async(execute_task, args=[key, dsk[key], data, queue,
-                                            get_id, raise_on_exception])
+            apply_async(execute_task,
+                        args=(key, marshall((dsk[key], data)),
+                              marshall, unmarshall,
+                              get_id, raise_on_exception),
+                        callback=queue.put)
+
+        def fill_pool_with_ready_tasks():
+            while state['ready'] and len(state['running']) < num_workers:
+                fire_task()
 
         # Seed initial tasks into the thread pool
-        while state['ready'] and len(state['running']) < num_workers:
-            fire_task()
+        fill_pool_with_ready_tasks()
 
         # Main loop, wait on tasks to finish, insert new ones
         while state['waiting'] or state['ready'] or state['running']:
-            key, res, tb, worker_id = queue.get()
+            key, res_info = queue.get()
+            try:
+                res, tb, worker_id = unmarshall(res_info)
+            except Exception as e:
```

Comment:
It's in the main process, yes, but it comes about from normal execution.  These aren't errors that the user can do anything about or that they should really observe.  I think it's reasonable for us to shield them in these cases.

I don't recall the exact application, but this behavior arose from an explicit need.  People were doing something reasonable that raised some hard-to-serialized exception.  Dask failed to get that exception back to them.  We could give them something that looked very similar though, which was appreciated at the time.